File: 1508-longest-happy-prefix/longest-happy-prefix.py
class Solution:
    def longestPrefix(self, s: str) -> str:
        p_hash = 0
        s_hash = 0
        BASE = 31
        MOD = 10 ** 9 + 7
        power = 1
        max_len = 0
        n = len(s)
        for i in range(n - 1):
            p_hash = (p_hash * BASE + ord(s[i])) % MOD
            s_hash = (s_hash + ord(s[n - i - 1]) * power) % MOD
            power = power * BASE % MOD
            if p_hash == s_hash:
                max_len = i + 1
        return s[:max_len]

File: 1508-longest-happy-prefix/longest-happy-prefix.cpp
class Solution {
public:
    string longestPrefix(string s) {
        int base = 31;
        int longest_len = 0;
        long long mod = 10e9 + 7;
        long long power = 1;
        long long pre_hash = 0;
        long long suf_hash = 0;
        for (int i = 0; i < s.size() - 1; ++i) {
            pre_hash = (pre_hash * base + s[i]) % mod;
            suf_hash = (suf_hash + s[s.size() - i - 1] * power) % mod;
            power = (power * base) % mod;
            if (pre_hash == suf_hash) {
                longest_len = i + 1;
            }
        }
        return s.substr(0, longest_len);
    }
};